Penpa parser snapshot testing

nmay231 opened the pull request:
While these are not visual tests, it's at least a starting point for tests.

The basic strategy is to load the main page in an iframe run all the tests in the browser. Mocha has this functionality which is nice. However, since snapshot tests need some access to the filesystem to save the snapshots, there is a test server that accepts requests to save files. It's feels a bit janky, but it works. If you have a nicer strategy, I'm all ears :).

This is a draft PR because of a couple of things.

1. This only snapshots `pu.pu_q`, `pu.mode`, and `usertab_choices`.  What other info do you think should be added? I'm holding off on adding all of the testing links for now.
2. Do you still want to keep `TESTING.md`? The Penpa links could also have a section in the tests.

Some things for the future:

1. Run this in a headless browser to speed up tests. It's not at all necessary at the moment, but I'm sure doing all the DOM painting is not cheap.
2. The test page reloads if any of the tests change, but not if any part of the main page changes. It would be a nice QoL addition.
3. Visual testing with SVG screenshots.

Cheers

swaroopg92 commented:
I read your comments but haven't found time to give more thoughts on it, I think November would be pretty busy with the World Puzzle Convention coming and most likely I will resume things after that.
I think I am ok with any initial testing stuff for now, as long as you can provide a clear instructions in a document to-do testing :)
Let's leave TESTING.md as is for now, no harm in having a file with links, we can remove it later once we have a good testing setup.

nmay231 commented:
Added the links from `TESTING.md` and fixed a couple of bugs. You can read `test/README.md` for details on running the tests. When you get time of course :)

swaroopg92 commented:
Hi, I was going through this,
I think using edit_url for testing might be a problem.
The reason is, as the features of Penpa expand, there will be more information added to the URL and so if we are generating a new edit_url to compare with the old URL that has been stored for testing purposes, they will never match directly.
I will have to give more thoughts on this.

nmay231 commented:
I hadn't thought of that... I've been working on the rest of the puzzlink parsers off and on. I'm thinking of just adding the rest in one PR (if that's not an issue). I could finish that up and then change this PR to snapshot only the properties we need instead of the whole edit_url.

In the meantime, here's a list of what we should probably include:

- cols, rows, grid type
- outside spacing (nb_space1-4)
- `pu.pu_q`, `pu.pu_a`
- `pu.mode.qa`
- penpalite selection
- The whole `pu.mode`?

swaroopg92 commented:
I will think about the list. But sounds good with single PR with all puzzlink stuff.

nmay231 commented:
Closing in favor of #68
